lib/engine-jit/src/artifact.rs

```diff
@@ -325,15 +327,32 @@ impl Artifact for JITArtifact {
         &self.func_data_registry
     }
     fn serialize(&self) -> Result<Vec<u8>, SerializeError> {
-        // let mut s = flexbuffers::FlexbufferSerializer::new();
-        // self.serializable.serialize(&mut s).map_err(|e| SerializeError::Generic(format!("{:?}", e)));
-        // Ok(s.take_buffer())
-        let bytes = bincode::serialize(&self.serializable)
-            .map_err(|e| SerializeError::Generic(format!("{:?}", e)))?;
-
         // Prepend the header.
         let mut serialized = Self::MAGIC_HEADER.to_vec();
-        serialized.extend(bytes);
+
+        let mut metadata_binary = vec![0; 12];
+        let mut writable = &mut metadata_binary[..];
+        let serialized_data = self.serializable.serialize()?;
+        let length = serialized_data.len();
+        leb128::write::unsigned(&mut writable, length as u64).expect("Should write number");
+
+        serialized.extend(metadata_binary);
+
+        let align = std::mem::align_of::<SerializableModule>() as u64;
+        // println!("[SER] Metadata: First 16 bytes {:?}.\nLast 16 bytes: {:?}\nTotal length: {}", &serialized_data[..16], &serialized_data[length-16..], length);
+        let (serialized, offset) = append_data(serialized, &serialized_data, align);
+        // println!("[SER] First 50 bytes {:?}.\nAlign: {} - Offset: {}\nLast 50 bytes: {:?}", &serialized[..50], align, offset, &serialized[serialized.len()-50..]);
         Ok(serialized)
     }
 }
+
+pub fn append_data(mut prev_data: Vec<u8>, data: &[u8], align: u64) -> (Vec<u8>, u64) {
+    let align = align as usize;
```

Comment:
There's not much reason to use `u64` over `usize` in this function, if we're talking about offsets into an array, `usize` is the most natural type

Reply:
`u64` will shield us in `32` bit systems?

Reply:
Because it's referring to an offset in memory it doesn't matter. On a 32bit system you only have 32 bits of addressable memory. A `usize` is defined to be the size big enough to cover all addressable memory, so when dealing with offsets it's a natural size. It makes sense to cast it to a u64 before writing it if you want it to take up a constant 8 bytes on all platforms, but when dealing with pointers / offsets, a `usize` will always be good enough and do the right thing on all platforms.